### ISSUE TITLE
Failed to install on macOS

### ISSUE THREAD

**Issue by phuclv90** [bkerler/edl#547]
I've followed the steps on the main readme file but it always failed at this point

```
Installed /opt/homebrew/lib/python3.12/site-packages/edlclient-3.62-py3.12.egg
Processing dependencies for edlclient==3.62
Searching for pylzma
Reading https://pypi.org/simple/pylzma/
Downloading https://files.pythonhosted.org/packages/03/d8/10ef072c3cd4301a65a1b762b09eefa02baf8da23b9ea77ebe9546499975/pylzma-0.5.0.tar.gz#sha256=b874172afbf37770e643bf2dc9d9b6b03eb95d8f8162e157145b3fe9e1b68a1c
Best match: pylzma 0.5.0
Processing pylzma-0.5.0.tar.gz
Writing /tmp/easy_install-fmj7cduw/pylzma-0.5.0/setup.cfg
Running pylzma-0.5.0/setup.py -q bdist_egg --dist-dir /tmp/easy_install-fmj7cduw/pylzma-0.5.0/egg-dist-tmp-76y3p9x_
no previously-included directories found matching 'src/sdk.orig'
/opt/homebrew/lib/python3.12/site-packages/setuptools/_distutils/cmd.py:66: SetuptoolsDeprecationWarning: setup.py install is deprecated.
!!

        ********************************************************************************
        Please avoid running ``setup.py`` directly.
        Instead, use pypa/build, pypa/installer or other
        standards-based tools.

        See https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html for details.
        ********************************************************************************

!!
  self.initialize_options()
/tmp/easy_install-fmj7cduw/pylzma-0.5.0/setup.py:105: UnsupportedPlatformWarning: Multithreading is not supported on the platform "darwin",
please contact mail@joachim-bauch.de for more informations.
  warn("""\
src/pylzma/pylzma.c:296:5: warning: 'PyEval_InitThreads' is deprecated [-Wdeprecated-declarations]
    PyEval_InitThreads();
    ^
/opt/homebrew/opt/python@3.12/Frameworks/Python.framework/Versions/3.12/include/python3.12/ceval.h:132:1: note: 'PyEval_InitThreads' has been explicitly marked deprecated here
Py_DEPRECATED(3.9) PyAPI_FUNC(void) PyEval_InitThreads(void);
^
/opt/homebrew/opt/python@3.12/Frameworks/Python.framework/Versions/3.12/include/python3.12/pyport.h:317:54: note: expanded from macro 'Py_DEPRECATED'
#define Py_DEPRECATED(VERSION_UNUSED) __attribute__((__deprecated__))
                                                     ^
1 warning generated.
src/pylzma/pylzma_aes.c:158:5: error: incompatible pointer to integer conversion initializing 'Py_ssize_t' (aka 'long') with an expression of type 'void *' [-Wint-conversion]
    NULL,                                /* printfunc  tp_print;   */
    ^~~~
/Library/Developer/CommandLineTools/usr/lib/clang/15.0.0/include/stddef.h:89:16: note: expanded from macro 'NULL'
#  define NULL ((void*)0)
               ^~~~~~~~~~
1 error generated.
```

In fact the first error happened was the below. And I managed to passed that with `brew install python-setuptools` but in the end the above error occured
```
src/edl/setup.py", line 2, in <module>
    from setuptools import setup, find_packages
ModuleNotFoundError: No module named 'setuptools'
```

**Comment by mrlambchop** [comment]
Same issue for me. Mac OS  Sonoma 14.2 (23C64)

**Comment by rstremke** [comment]
Same here. Running 14.4 Beta (23E5205c).

**Comment by clarkewing** [comment]
Same issue here. maxOS 14.4.1 (23E224)

**Comment by fridaynext** [comment]
Stack Overflow to the rescue!

https://stackoverflow.com/questions/78282287/why-am-i-getting-incompatible-pointer-to-integer-conversion-initializing-py-ssi

I had to also uninstall Xcode from my system, since it was already updated to 15.3.

## **tl;dr:**
- Downgrade Xcode Command Line Tools to 15.1
- uninstall Xcode 15.3 if you have it installed
- install should now work (did for me!)

**Comment by phuclv90** [comment]
I don't have Xcode as well as `/Library/Devloper/CommandLineTools` in my case so it probably uses the default `/usr/bin/cc`

**Comment by JohnnyPrimus** [comment]
dont use 'sudo python3 setup.py install'

instead do 'sudo pip3 install edl'

**Comment by randolph-** [comment]
Alternatively, you can try downgrading `clang` temporarily without having to fuss with Xcode.

```sh
$ which clang
/usr/bin/clang

$ clang -v
Apple clang version 15.0.0 (clang-1500.3.9.4)
Target: arm64-apple-darwin23.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

$ brew install llvm@14
$ export PATH="/opt/homebrew/opt/llvm@14/bin:$PATH"

$ which clang
/opt/homebrew/opt/llvm@14/bin/clang

$ clang -v
Homebrew clang version 14.0.6
Target: arm64-apple-darwin23.4.0
Thread model: posix
InstalledDir: /opt/homebrew/opt/llvm@14/bin
```

Afterwards, the build and install work as expected for me. Following the install, you can run `brew uninstall llvm@14` to uninstall, or simply restart your terminal session to restore your `PATH` while keeping `clang-14` available should you need it down the line.

**Comment by phuclv90** [comment]
thanks @randolph- I was able to build. But now I'm seeing another issue https://github.com/bkerler/edl/issues/556

**Comment by rmohns** [comment]
> Alternatively, you can try downgrading `clang` temporarily without having to fuss with Xcode.
> (snip instructions)

***Thank you!*** This is much faster than downgrading all of Xcode! You're a 🌟!

**Comment by BlairSadewitz** [comment]
With this [patch](https://github.com/fancycode/pylzma/pull/82), pylzma builds.

-rw-r--r--  1 blair  staff  156705 Oct  3 19:30 pylzma-v0.5.0_18_g2fe0dev-cp39-cp39-macosx_10_9_universal2.whl

**Comment by smiba** [comment]
You can use it without pylzma I think

I managed to get it working by doing the following:
```
cd ~\Downloads
git clone https://github.com/bkerler/edl
cd edl
git submodule update --init --recursive
brew install libusb git python-setuptools
#The following is bad, but I hate python and I can't bother figuring out virtual environments, run at your own risk:
pip3 install -r requirements.txt  --break-system-packages
#USER ACTION: Before continuing remove pylzma from the setup.py file
#nano setup.py
sudo python3 setup.py install
```

Dumping an eMMC NAND as I write this

**Comment by BlairSadewitz** [comment]
```
> #The following is bad, but I hate python and I can't bother figuring out virtual environments, run at your own risk:
```
Lol, yeah, all that drives me crazy too, sometimes, but I'm not much of a programmer, so I just go with it.  Well, that and micromamba.  So much faster.

I kinda rely on python to do things tho so that whole --break-system-packages thing probably isn't for me.  Not even being sarcastic.

**Comment by fivestones** [comment]
The patch from @BlairSadewitz worked for me.
My full install was:
```
brew install libusb git
git clone http://github.com/bkerler/edl.git
cd edl
git submodule update --init --recursive
python3 setup.py build
pip install --upgrade pip wheel
pip install "git+https://github.com/rickmark/pylzma@struct_fix"
sudo python3 setup.py install
```

**Comment by Niek** [comment]
I personally prefer `uv` instead, so it doesn't even require a system-wide install:
```bash
brew install libusb # git is already installed if you have `brew` :)
git clone http://github.com/bkerler/edl.git
cd edl
git submodule update --init --recursive
uv run --with setuptools setup.py build
uv run --with docopt --with pyserial --with pyusb --with lxml --with pycryptodome --with colorama --with exscript --with requests --with passlib ./edl
```
